### PR TITLE
Posterior Hybrid Bayes Net

### PR DESCRIPTION
**Pull request by varunagrawal**
This PR updates the incremental hybrid elimination to build up the full posterior bayes net.

@dellaert we would require some feedback from you as well here. :)

### CODE REVIEW

**Review by ProfFan — APPROVED** [review]

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 1]
```diff
@@ -32,23 +33,30 @@ void gtsam::IncrementalHybrid::update(gtsam::GaussianHybridFactorGraph graph,
...
+                         conditional)) {
             graph.push_back(df);
           }
           break;
         }
       }
     }
+  } else {
+    // Initialize an empty HybridBayesNet
+    hybridBayesNet_ = boost::make_shared<HybridBayesNet>();
```
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:49]: Why a shared pointer? In another PR I am trying to eradicate all shared pointers to what are essentially vectors. Actually, not looking for a reason, but a change :-)

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 2]
```diff
@@ -32,23 +33,30 @@ void gtsam::IncrementalHybrid::update(gtsam::GaussianHybridFactorGraph graph,
...
   }
 
   // Eliminate partially.
-  std::tie(hybridBayesNet_, remainingFactorGraph_) =
+  HybridBayesNet::shared_ptr bayesNetFragment;
```
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:53]: will soon no longer be a shared pointer, but that will be for later

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 3]
```diff
@@ -32,23 +33,30 @@ void gtsam::IncrementalHybrid::update(gtsam::GaussianHybridFactorGraph graph,
...
+  std::tie(bayesNetFragment, remainingFactorGraph_) =
       graph.eliminatePartialSequential(ordering);
 
+  // Add the partial bayes net to the posterior bayes net.
+  hybridBayesNet_->push_back<HybridBayesNet>(*bayesNetFragment);
```
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:58]: hybridBayesNet_.push_back

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 4]
```diff
@@ -62,17 +70,15 @@ void gtsam::IncrementalHybrid::update(gtsam::GaussianHybridFactorGraph graph,
     // Let's assume that the structure of the last discrete density will be the
     // same as the last continuous
     std::vector<double> probabilities;
-    // TODO(fan): The number of probabilities can be lower than the actual number of choices
-    discreteFactor->visit([&](const double &prob) {
-      probabilities.emplace_back(prob);
-    });
+    // TODO(fan): The number of probabilities can be lower than the actual
+    // number of choices
```
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.cpp:74]: You know that half of what I do in this review is trying to figure out whether it is just spurious reformatting?

File: gtsam/hybrid/IncrementalHybrid.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/hybrid/IncrementalHybrid.cpp:49]: PS, line 29 would have to be changed as well -> just check if empty

**Review by dellaert — APPROVED** [review]
Please don't re-format, or re-format in a different PR.

**Comment by dellaert** [comment]
About re-formatting: this is for future reference. You make the life of a reviewer very hard.

**Comment by varunagrawal** [comment]
Ouch, sorry about that. The amount of formatting changes seemed small enough to not warrant a separate PR. :(

**Comment by varunagrawal** [comment]
So the main review comments seem to be about the shared_ptr. I'll make that update in this PR itself.

**Comment by dellaert** [comment]
> Ouch, sorry about that. The amount of formatting changes seemed small enough to not warrant a separate PR. :(

The issue is not about a separate PR, but the difficulty in reviewing. every diff has to be carefully examined as to whether is actually a change or not.

**Comment by varunagrawal** [comment]
You're correct. I need to do better...

**Comment by varunagrawal** [comment]
Also, I'll make the change from shared pointer to object in a different PR since there is some considerable change happening.

**Pull request merged**